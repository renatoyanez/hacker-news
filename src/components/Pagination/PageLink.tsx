import { HTMLProps } from 'react';
import { Typography } from 'components';
import styles from './pageLink.module.css';

export type Props = HTMLProps<HTMLAnchorElement> & {
	active?: boolean;
};

export default function PageLink({
	active,
	disabled,
	children,
	...otherProps
}: Props) {
	if (disabled) {
		return (
			<span
				className={`${styles.pageLink} ${
					active ? styles.active : ''
				}`}
			>
				<Typography type={active ? 'textStyle7' : 'textStyle3'}>
					{children}
				</Typography>
			</span>
		);
	}

	return (
		<a
			className={`${styles.pageLink} ${active ? styles.active : ''}`}
			aria-current={active ? 'page' : undefined}
			{...otherProps}
		>
			<Typography type={active ? 'textStyle7' : 'textStyle3'}>
				{children}
			</Typography>
		</a>
	);
}
