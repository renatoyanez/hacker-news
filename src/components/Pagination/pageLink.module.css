.pageLink {
	position: relative;
	display: inline-flex;
	border-radius: 60%;
	background-color: #ffffff;
	padding: 10px 12px;
	text-decoration: none;
	cursor: pointer;
	margin: 0 3px;
}

.pageLink:first-child {
	border-radius: 1px;
}

.pageLink:last-child {
	border-radius: 1px;
}

.pageLink:not(:first-child) {
	margin-left: -1px;
}

.pageLink:hover,
.pageLink:focus {
	color: #0a58ca;
	background-color: #e9ecef;
}

.pageLink:focus {
	z-index: 3;
}

.active {
	z-index: 2;
	color: #ffffff;
	border: 1px solid #0d6efd;
	background-color: #0d6efd;
}

.pageLink.disabled {
	color: #6c757d;
	pointer-events: none;
}
