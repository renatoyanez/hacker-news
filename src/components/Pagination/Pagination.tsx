import { getPaginationItems } from './helper';
import PageLink from './PageLink';
import styles from './pagination.module.css';

export interface PaginationProps {
	currentPage: number;
	lastPage: number;
	maxLength: number;
	setCurrentPage: (page: number) => void;
}

const Pagination = ({
	currentPage,
	lastPage,
	maxLength,
	setCurrentPage,
}: PaginationProps) => {
	const pageNums = getPaginationItems(
		currentPage,
		lastPage,
		maxLength,
	);

	const correctedNums = pageNums.map(num => Math.ceil(num));

	return (
		<nav className={styles.pagination} aria-label="Pagination">
			<PageLink
				disabled={currentPage === 1}
				onClick={() => setCurrentPage(currentPage - 1)}
			>
				Previous
			</PageLink>
			{correctedNums.map((pageNum, index) => {
				return (
					<PageLink
						key={index}
						data-testid="number-of-page"
						active={currentPage === pageNum}
						disabled={isNaN(pageNum)}
						onClick={() => setCurrentPage(pageNum)}
					>
						{!isNaN(pageNum) ? pageNum : '...'}
					</PageLink>
				);
			})}
			<PageLink
				disabled={currentPage === lastPage}
				onClick={() => setCurrentPage(currentPage + 1)}
			>
				Next
			</PageLink>
		</nav>
	);
};

export default Pagination;
