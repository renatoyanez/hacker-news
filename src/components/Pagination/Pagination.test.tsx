import React from 'react';
import { create } from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import Pagination, { PaginationProps } from './Pagination';

const props: PaginationProps = {
	currentPage: 1,
	lastPage: 9,
	maxLength: 9,
	setCurrentPage: jest.fn(),
};

describe('Pagination', () => {
	it('renders correctly', () => {
		const tree = create(<Pagination {...props} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('calls on change when clicked', () => {
		const mockClick = jest.fn();
		const { getByText, queryAllByTestId } = render(
			<Pagination {...props} setCurrentPage={mockClick} />,
		);
		const button = getByText(/1/i);
		fireEvent.click(button);

		const numberOfPage = queryAllByTestId(
			'number-of-page',
		) as Element[];

		expect(numberOfPage[0]).toHaveClass('pageLink active');

		expect(mockClick).toHaveBeenCalledTimes(1);
	});

	it('calls on change multiple times', () => {
		const mockClick = jest.fn();
		const component = render(
			<Pagination {...props} setCurrentPage={mockClick} />,
		);
		const button1 = component.getByText(/1/i);
		const button2 = component.getByText(/2/i);
		const button3 = component.getByText(/3/i);
		const button4 = component.getByText(/4/i);

		fireEvent.click(button1);
		fireEvent.click(button3);
		fireEvent.click(button2);
		fireEvent.click(button4);

		expect(mockClick).toHaveBeenCalledTimes(4);
	});
});
