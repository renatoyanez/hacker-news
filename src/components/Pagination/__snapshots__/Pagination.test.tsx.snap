// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<nav
  aria-label="Pagination"
  className="pagination"
>
  <span
    className="pageLink "
  >
    <p
      className="textStyle3"
    >
      Previous
    </p>
  </span>
  <a
    aria-current="page"
    className="pageLink active"
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle7"
    >
      1
    </p>
  </a>
  <a
    className="pageLink "
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      2
    </p>
  </a>
  <a
    className="pageLink "
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      3
    </p>
  </a>
  <a
    className="pageLink "
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      4
    </p>
  </a>
  <a
    className="pageLink "
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      5
    </p>
  </a>
  <a
    className="pageLink "
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      6
    </p>
  </a>
  <a
    className="pageLink "
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      7
    </p>
  </a>
  <a
    className="pageLink "
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      8
    </p>
  </a>
  <a
    className="pageLink "
    data-testid="number-of-page"
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      9
    </p>
  </a>
  <a
    className="pageLink "
    onClick={[Function]}
  >
    <p
      className="textStyle3"
    >
      Next
    </p>
  </a>
</nav>
`;
